### ISSUE TITLE
[DataGrid] Allow sortable indicator on column headers without a sort applied

### ISSUE THREAD

**Issue by emily-ellevation** [mui-org/material-ui-x#1076]
When a sortable column is not sorted is there any way to surface a 3rd icon type with the existing API? As it stands, it appears the sort icon is removed from the DOM altogether when a column is unsorted.

**3rd Type: Sortable but no sort applied:**
<img width="22" alt="Screen Shot 2021-02-18 at 2 47 47 PM" src="https://user-images.githubusercontent.com/52083969/108431866-53ff4b00-71f8-11eb-966e-9daa0ee56da6.png">
Ascending:
<img width="20" alt="Screen Shot 2021-02-18 at 2 47 37 PM" src="https://user-images.githubusercontent.com/52083969/108431849-4ea20080-71f8-11eb-875f-5ae15d4e8b9c.png">
Descending:
<img width="22" alt="Screen Shot 2021-02-18 at 2 47 56 PM" src="https://user-images.githubusercontent.com/52083969/108431860-52358780-71f8-11eb-9df8-0b4311c18c77.png">

- [X] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Columns that have `sortable:true` but have a 'null' sort direction can display a 3rd neutral sort icon to indicate they are sortable.

## Examples 🌈

In the following example, the table is sorted by Student Name, School has no sort applied but displays a neutral 'unsorted/sortable' indicator.

<img width="342" alt="Screen Shot 2021-02-18 at 2 42 29 PM" src="https://user-images.githubusercontent.com/52083969/108431350-9b390c00-71f7-11eb-87e7-432fc1b99638.png">

## Motivation 🔦

We focus a lot on explicit discoverability in our UIs. Unsorted columns having no indicator that they are sortable does not allow us to give explicit feedback that columns _can_ be sorted and causes users not to use sorting. I'm aware I can apply a default sort with sort model, but this does not guarantee users can discover that _other_ columns are sortable.

**Comment by dtassone** [comment]
Good point. ATM the null sort has no icon but yes we should allow a custom icon in components to override that behavior.

**Comment by emily-ellevation** [comment]
Awesome, glad to hear. Do you have a sense of priority/timing? Is this something I could possibly take a shot at as a first issue?

**Comment by oliviertassinari** [comment]
@emily-ellevation If you have a look, it would be great. At the very least, I would start cleanup the code. We don't need the extra div nor the duplication of the icon button:

```diff
diff --git a/packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx b/packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx
index 252b9aa5..711bd920 100644
--- a/packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx
+++ b/packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx
@@ -27,30 +27,24 @@ export const GridColumnHeaderSortIcon = React.memo(function GridColumnHeaderSort
     return null;
   }

+  const iconButton = (
+    <IconButton
+      aria-label={apiRef!.current.getLocaleText('columnHeaderSortIconLabel')}
+      title={apiRef!.current.getLocaleText('columnHeaderSortIconLabel')}
+      size="small"
+    >
+      {getIcon(apiRef!.current.components, direction)}
+    </IconButton>
+  );
+
   return (
     <div className="MuiDataGrid-iconButtonContainer">
-      <div>
-        {index != null && (
-          <Badge badgeContent={index} color="default">
-            <IconButton
-              aria-label={apiRef!.current.getLocaleText('columnHeaderSortIconLabel')}
-              title={apiRef!.current.getLocaleText('columnHeaderSortIconLabel')}
-              size="small"
-            >
-              {getIcon(apiRef!.current.components, direction)}
-            </IconButton>
-          </Badge>
-        )}
-        {index == null && (
-          <IconButton
-            aria-label={apiRef!.current.getLocaleText('columnHeaderSortIconLabel')}
-            title={apiRef!.current.getLocaleText('columnHeaderSortIconLabel')}
-            size="small"
-          >
-            {getIcon(apiRef!.current.components, direction)}
-          </IconButton>
-        )}
-      </div>
+      {index != null && (
+        <Badge badgeContent={index} color="default">
+          {iconButton}
+        </Badge>
+      )}
+      {index == null && iconButton}
     </div>
   );
 });
```

Then, I would fix the overflow issue: 

<img width="137" alt="Screenshot 2021-02-21 at 19 20 02" src="https://user-images.githubusercontent.com/3165635/108634421-cfbbfa80-7479-11eb-887f-2d07ecdf3a23.png">

```diff
diff --git a/packages/grid/_modules_/grid/components/containers/GridRootStyles.ts b/packages/grid/_modules_/grid/components/containers/GridRootStyl
es.ts
index 3b7db046..a594f24c 100644
--- a/packages/grid/_modules_/grid/components/containers/GridRootStyles.ts
+++ b/packages/grid/_modules_/grid/components/containers/GridRootStyles.ts
@@ -103,10 +103,8 @@ export const useStyles = makeStyles(
           alignItems: 'center',
         },
         '& .MuiDataGrid-colCellTitleContainer': {
-          textOverflow: 'ellipsis',
-          overflow: 'hidden',
-          whiteSpace: 'nowrap',
-          display: 'inline-flex',
+          display: 'flex',
+          minWidth: 0, // Allow the overflow ellipsis to trigger
           flex: 1,
         },
         '& .MuiDataGrid-colCellNumeric .MuiDataGrid-iconButtonContainer': {
@@ -276,8 +274,8 @@ export const useStyles = makeStyles(
         },
         '& .MuiDataGrid-colCell-draggable': {
           display: 'flex',
-          width: '100%',
-          justifyContent: 'inherit',
+          minWidth: 0, // Allow the overflow ellipsis to trigger
+          flex: 1,
         },
         '& .MuiDataGrid-colCell-dragging': {
           background: theme.palette.background.paper,
```

---

So regarding fixing this problem. I believe it will require some changes in the API. Maybe:

- Make `hideSortIcons` an enum, always visible, always hidden, only if active
- Add a new Icon customization key

**Comment by emily-ellevation** [comment]
I got pulled away but thank you very much for the detailed overview @oliviertassinari. Will update here if I'm able to pick this up, this feature is definitely still a need/on our radar just gotta find the time

**Comment by oliviertassinari** [comment]
@emily-ellevation Ok, if you have some time, starting with the cleanup of the code would already help.

**Comment by oliviertassinari** [comment]
On this topic of affordance with the sortable feature, the data grid handles the problem differently to the table component. See:

![sorting](https://user-images.githubusercontent.com/3165635/115784715-c6d69f00-a3be-11eb-9d80-d1b6a1886e84.gif)

Should we open a new issue for it?

**Comment by m4theushw** [comment]
We could achieve this same behavior by using the up arrow as the unsorted icon and hiding it with CSS. It could be part of #1415.

**Comment by dtassone** [comment]
> We could achieve this same behavior by using the up arrow as the unsorted icon and hiding it with CSS. It could be part of #1415.

We need to respect the order of the `sortingOrder ` prop

**Comment by emily-ellevation** [comment]
Our designers are over the moon! Thank you so much 🥳

**Comment by WilliamZimmermann** [comment]
> * Make `hideSortIcons` an enum, always visible, always hidden, only if active

This request was closed but was it implemented? Would be nice (and very useful) to display the sort icon even if the column is not sorted. The user should be able to see (without the need to pass the mouse over each column) witch columns are sortable and with columns are not.

**Comment by oliviertassinari** [comment]
> This request was closed but was it implemented? 

@WilliamZimmermann Part of the pain was solved in #1415. For example, you can do: https://codesandbox.io/s/basicsortinggrid-material-demo-forked-z9zdd?file=/demo.js:0-1140

<img width="432" alt="Screenshot 2021-09-25 at 00 04 02" src="https://user-images.githubusercontent.com/3165635/134744280-8d465dfc-cee8-4313-af5a-44d77e3c5533.png">

```jsx
import * as React from "react";
import { DataGrid } from "@mui/x-data-grid";
import { useDemoData } from "@mui/x-data-grid-generator";
import ArrowDropDownIcon from "@mui/icons-material/ArrowDropDown";
import Box from "@mui/material/Box";

export default function BasicSortingGrid() {
  const { data } = useDemoData({
    dataSet: "Commodity",
    rowLength: 10,
    maxColumns: 6
  });

  const [sortModel, setSortModel] = React.useState([
    {
      field: "commodity",
      sort: "asc"
    }
  ]);

  return (
    <Box
      style={{ height: 400, width: "100%" }}
      sx={{
        "& .MuiDataGrid-root .MuiDataGrid-columnHeader .MuiDataGrid-iconButtonContainer": {
          width: "auto",
          visibility: "visible"
        },
        "& .MuiDataGrid-root .MuiDataGrid-columnHeader:not(.MuiDataGrid-columnHeader--sorted) .MuiDataGrid-sortIcon": {
          opacity: 0.5
        }
      }}
    >
      <DataGrid
        {...data}
        sortModel={sortModel}
        components={{
          ColumnUnsortedIcon: ArrowDropDownIcon
        }}
        onSortModelChange={(model) => setSortModel(model)}
      />
    </Box>
  );
}
```

IMHO, we should follow up with updating https://mui.com/components/data-grid/style/#custom-theme to match the icons of https://ant.design/components/table/#components-table-demo-head.

We could also consider a dedicated header in https://mui.com/components/data-grid/sorting/.
